frontend/option_def.py

```diff
@@ -150,6 +150,7 @@ def DoneWithImplementedOptions(self):
 _BASIC_PARSE_OPTIONS = [
     'parse_at',  # @foo, @array(a, b)
     'parse_proc',  # proc p { ... }
+    'parse_func',  # func f(x) { ... }
```

Comment:
Enabling `parse_func` in OSH is tricky because of the following:

```bash
func() { echo } # valid
func f { echo } # valid
func f() { echo } # invalid, this is where we add funcs into the grammar
```

So to parse YSH funcs, without breaking the first two cases, we'd have to do some fancier look-ahead.

Now I don't think the first two lines are a common case, which is why I didn't add the necessary lookahead in func parsing to support them. Then we can keep the parser fast and errors sane. But this comes at the cost of making `parse_func` a breaking change.

Btw, procs have the same issue:

```
$ bin/osh -c 'proc() { echo; }'; echo $?
0
$ bin/ysh -c 'proc() { echo; }'; echo $?
  proc() { echo; }
       ^
[ -c flag ]:1: Invalid proc name ')'
2
```